libass/ass.c

```diff
@@ -1535,17 +1535,33 @@ ASS_Track *ass_new_track(ASS_Library *library)
 
 int ass_track_set_feature(ASS_Track *track, ASS_Feature feature, int enable)
 {
-    switch (feature) {
-    case ASS_FEATURE_INCOMPATIBLE_EXTENSIONS:
-        //-fallthrough
+    if (feature >= sizeof(track->parser_priv->feature_flags) * CHAR_BIT || feature < 0)
+        return -1;
+
+    // all supported non-meta features
+    static const uint32_t supported = 0
 #ifdef USE_FRIBIDI_EX_API
-    case ASS_FEATURE_BIDI_BRACKETS:
-        track->parser_priv->bidi_brackets = !!enable;
+        | FEATURE_MASK(ASS_FEATURE_BIDI_BRACKETS)
 #endif
-        return 0;
+        ;
+    uint32_t requested = 0;
+
+    switch (feature) {
+    case ASS_FEATURE_INCOMPATIBLE_EXTENSIONS:
+        requested = supported;
+        break;
     default:
-        return -1;
+        if (!(FEATURE_MASK(feature) & supported))
+            return -1;
+        requested = FEATURE_MASK(feature);
     }
+
+    if (enable)
+        track->parser_priv->feature_flags |= requested;
+    else
+        track->parser_priv->feature_flags &= ~(requested);
```

Comment:
Is there any meaning in additional brackets?

Reply:
No, probably a leftover from the macro-version; removed them now.